File: networking/static_config.py
"""Builders for the browser-only handshake payloads (ai_choices + server_config).

These snapshot the Python-side constants and AI registry into plain JSON-able
dicts so the browser client doesn't have to duplicate ``config/`` in
TypeScript. They are built once by callers that already own a registry
(``server.py``, ``app.py``, ``internal_server.py``) and passed into
``GameHost`` — ``host.py`` itself never imports ``config`` or the registry,
keeping the network thread free of that import cost.

Desktop (TCP) clients ignore these extra messages; only the browser client
reads them.
"""
from __future__ import annotations


def build_ai_choices(registry) -> dict:
    """Snapshot the AI registry for the lobby AI dropdowns.

    The browser can't enumerate Python AI classes, so this list is the only
    way it learns which bots exist.
    """
    choices = registry.get_choices(include_deprecated=True)
    deprecated = registry.get_deprecated_ids()
    return {
        "choices": [[ai_id, ai_name] for ai_id, ai_name in choices],
        "deprecated": sorted(deprecated),
    }


def build_static_config() -> dict:
    """Snapshot the display constants the browser client needs to render and
    label the game. Sent live each connect, so it stays in sync with ``config/``
    as the game is balanced (no hand-copied constants in TypeScript)."""
    from config import settings as s
    from config.unit_types import UNIT_TYPES, T2_NAMES, get_spawnable_types
    from systems.replay import RECORD_INTERVAL
    from systems.abilities import ABILITY_REGISTRY

    return {
        "unit_types": UNIT_TYPES,
        "t2_names": T2_NAMES,
        "spawnable_types": list(get_spawnable_types().keys()),
        "player_colors": [list(c) for c in s.PLAYER_COLORS],
        "team_colors": {str(t): list(c) for t, c in s.TEAM_COLORS.items()},
        "selected_color": list(s.SELECTED_COLOR),
        "broadcast_interval": RECORD_INTERVAL,
        "timing": {
            "tick_rate": s.TICK_RATE,
            "fixed_dt": s.FIXED_DT,
            "max_frame_dt": s.MAX_FRAME_DT,
        },
        "command_center": {
            "hp": s.CC_HP,
            "radius": s.CC_RADIUS,
            "spawn_interval": s.CC_SPAWN_INTERVAL,
            "spawn_range": s.CC_SPAWN_RANGE,
            "laser_range": s.CC_LASER_RANGE,
        },
        "metal": {
            "spot_radius": s.METAL_SPOT_RADIUS,
            "capture_radius": s.METAL_SPOT_CAPTURE_RADIUS,
            "extractor_radius": s.METAL_EXTRACTOR_RADIUS,
            "extractor_spawn_bonus": s.METAL_EXTRACTOR_SPAWN_BONUS,
            "reinforce_max_stacks": s.REINFORCE_MAX_STACKS,
        },
        "upgrades": {
            "outpost_duration": s.OUTPOST_UPGRADE_DURATION,
            "research_lab_duration": s.RESEARCH_LAB_UPGRADE_DURATION,
            "t2_spawn_bonus": s.T2_SPAWN_BONUS,
            # Stat details for the browser's upgrade tooltips (gui.py parity).
            "outpost": {
                "hp_bonus": s.OUTPOST_HP_BONUS,
                "heal_per_sec": s.OUTPOST_HEAL_PER_SEC,
                "los": s.OUTPOST_LOS,
                "laser_damage": s.OUTPOST_LASER_DAMAGE,
                "laser_range": s.OUTPOST_LASER_RANGE,
                "laser_cooldown": s.OUTPOST_LASER_COOLDOWN,
            },
            "research_lab": {
                "hp_bonus": s.RESEARCH_LAB_HP_BONUS,
            },
            "reinforce": {
                "max_stacks": s.REINFORCE_MAX_STACKS,
                "stack_interval": s.REINFORCE_STACK_INTERVAL,
                "hp_bonus": s.REINFORCE_HP_BONUS,
                "bonus_multiplier": s.REINFORCE_BONUS_MULTIPLIER,
            },
        },
        # name -> description for the HUD abilities panel (registry lives in
        # Python; the browser can't import systems/abilities.py).
        "ability_descriptions": {
            name: getattr(cls, "description", "")
            for name, cls in ABILITY_REGISTRY.items()
        },
        "camera": {
            "zoom_step": s.CAMERA_ZOOM_STEP,
            "max_zoom": s.CAMERA_MAX_ZOOM,
            "edge_pan_margin": s.EDGE_PAN_MARGIN,
            "edge_pan_speed": s.EDGE_PAN_SPEED,
        },
        "health_bar": {
            "width": s.HEALTH_BAR_WIDTH,
            "height": s.HEALTH_BAR_HEIGHT,
            "offset": s.HEALTH_BAR_OFFSET,
        },
    }

File: web/src/config/StaticConfig.ts
// Typed view of the server's `server_config` message (built by
// networking/static_config.py). Keeps the browser's display constants in sync
// with the Python `config/` without hand-copying them. Falls back to sane
// defaults if a field is missing so rendering never crashes.

import type { RGB } from "../ui/theme";

export interface UnitStat {
  hp: number;
  speed: number;
  radius: number;
  symbol: [number, number][] | null;
  can_attack: boolean;
  fov: number;
  turn_rate: number;
  los: number;
  hollow?: boolean;
  is_building?: boolean;
  is_t2?: boolean;
  spawn_count?: number;
  weapon?: Record<string, unknown>;
}

export interface OutpostStats {
  hp_bonus: number;
  heal_per_sec: number;
  los: number;
  laser_damage: number;
  laser_range: number;
  laser_cooldown: number;
}

export interface ReinforceStats {
  max_stacks: number;
  stack_interval: number;
  hp_bonus: number;
  bonus_multiplier: number;
}

export class StaticConfig {
  raw: Record<string, unknown>;
  unitTypes: Record<string, UnitStat>;
  t2Names: Record<string, string>;
  spawnableTypes: string[];
  playerColors: RGB[];
  teamColors: Record<number, RGB>;
  selectedColor: RGB;
  broadcastInterval: number;
  fixedDt: number;
  ccRadius: number;
  ccHp: number;
  ccLaserRange: number;
  outpostLaserRange: number;
  metalCaptureRadius: number;
  metalExtractorHp: number;
  outpostLos: number;
  zoomStep: number;
  maxZoom: number;
  edgePanMargin: number;
  edgePanSpeed: number;
  healthBarWidth: number;
  healthBarHeight: number;
  healthBarOffset: number;
  outpostUpgradeDuration: number;
  researchLabUpgradeDuration: number;
  t2SpawnBonus: number;
  outpost: OutpostStats;
  researchLabHpBonus: number;
  reinforce: ReinforceStats;
  abilityDescriptions: Record<string, string>;

  constructor(raw: Record<string, unknown> | null) {
    this.raw = raw ?? {};
    const r = this.raw;
    this.unitTypes = (r.unit_types as Record<string, UnitStat>) ?? {};
    this.t2Names = (r.t2_names as Record<string, string>) ?? {};
    this.spawnableTypes = (r.spawnable_types as string[]) ?? [];
    this.playerColors = ((r.player_colors as number[][]) ?? []).map(
      (c) => [c[0], c[1], c[2]] as RGB,
    );
    this.teamColors = {};
    const tc = (r.team_colors as Record<string, number[]>) ?? {};
    for (const [k, v] of Object.entries(tc)) {
      this.teamColors[Number(k)] = [v[0], v[1], v[2]] as RGB;
    }
    const sc = (r.selected_color as number[]) ?? [0, 255, 100];
    this.selectedColor = [sc[0], sc[1], sc[2]];
    this.broadcastInterval = (r.broadcast_interval as number) ?? 6;
    const timing = (r.timing as Record<string, number>) ?? {};
    this.fixedDt = timing.fixed_dt ?? 1 / 60;
    const cc = (r.command_center as Record<string, number>) ?? {};
    this.ccRadius = cc.radius ?? 10;
    this.ccHp = cc.hp ?? 1000;
    this.ccLaserRange = cc.laser_range ?? 75;
    this.outpostLaserRange = 75;
    const metal = (r.metal as Record<string, number>) ?? {};
    this.metalCaptureRadius = metal.capture_radius ?? 15;
    this.metalExtractorHp = 200;
    const up = (r.upgrades as Record<string, unknown>) ?? {};
    this.outpostUpgradeDuration = (up.outpost_duration as number) ?? 30;
    this.researchLabUpgradeDuration = (up.research_lab_duration as number) ?? 60;
    this.t2SpawnBonus = (up.t2_spawn_bonus as number) ?? 0.2;
    const op = (up.outpost as Partial<OutpostStats>) ?? {};
    this.outpost = {
      hp_bonus: op.hp_bonus ?? 50,
      heal_per_sec: op.heal_per_sec ?? 1,
      los: op.los ?? 140,
      laser_damage: op.laser_damage ?? 15,
      laser_range: op.laser_range ?? 75,
      laser_cooldown: op.laser_cooldown ?? 2,
    };
    this.researchLabHpBonus =
      ((up.research_lab as Record<string, number>) ?? {}).hp_bonus ?? 100;
    const rf = (up.reinforce as Partial<ReinforceStats>) ?? {};
    this.reinforce = {
      max_stacks: rf.max_stacks ?? 4,
      stack_interval: rf.stack_interval ?? 15,
      hp_bonus: rf.hp_bonus ?? 100,
      bonus_multiplier: rf.bonus_multiplier ?? 2,
    };
    this.abilityDescriptions = (r.ability_descriptions as Record<string, string>) ?? {};
    this.outpostLos = this.outpost.los;
    this.outpostLaserRange = this.outpost.laser_range;
    const cam = (r.camera as Record<string, number>) ?? {};
    this.zoomStep = cam.zoom_step ?? 1.1;
    this.maxZoom = cam.max_zoom ?? 3.0;
    this.edgePanMargin = cam.edge_pan_margin ?? 10;
    this.edgePanSpeed = cam.edge_pan_speed ?? 500;
    const hb = (r.health_bar as Record<string, number>) ?? {};
    this.healthBarWidth = hb.width ?? 24;
    this.healthBarHeight = hb.height ?? 3;
    this.healthBarOffset = hb.offset ?? 4;
  }

  teamColor(team: number): RGB {
    return this.teamColors[team] ?? this.playerColors[(team - 1) % Math.max(1, this.playerColors.length)] ?? [255, 255, 255];
  }

  /** Live LOS for an entity: prefer server-sent los, else unit-type base. */
  losFor(ut: string, sentLos: number | undefined, upgradeState?: string): number {
    if (ut === "metal_extractor" && upgradeState === "outpost") return this.outpostLos;
    if (sentLos !== undefined) return sentLos;
    return this.unitTypes[ut]?.los ?? 100;
  }

  speedFor(ut: string): number {
    return this.unitTypes[ut]?.speed ?? 0;
  }
}
